diff --git a/WebIDL.py b/WebIDL.py
index c474a7ba227..a107248ef1f 100644
--- a/WebIDL.py
+++ b/WebIDL.py
@@ -2153,6 +2153,7 @@ class IDLInterface(IDLInterfaceOrNamespace):
             or identifier == "Serializable"
             or identifier == "Abstract"
             or identifier == "Inline"
+            or identifier == "Transferable"
         ):
             # Known extended attributes that do not take values
             if not attr.noArguments():
-- 
2.51.0

