--- js-test-pre.orig.js	2016-04-08 22:35:15.629226767 +0200
+++ js-test-pre.js	2016-04-08 22:43:11.906092062 +0200
@@ -71,11 +71,25 @@
   }
 }
 
-function reportTestResultsToHarness(success, msg) {
-  if (window.parent.webglTestHarness) {
-    window.parent.webglTestHarness.reportResults(window.location.pathname, success, msg);
+(function() {
+  var WPT_TEST_ID = 0;
+
+  // Store the current WPT test harness `test` function
+  // if found, since it's overriden by some tests.
+  var wpt_test = window.test;
+  var wpt_assert_true = window.assert_true;
+
+
+  window.reportTestResultsToHarness = function reportTestResultsToHarness(success, msg) {
+    if (window.parent.webglTestHarness) {
+      window.parent.webglTestHarness.reportResults(window.location.pathname, success, msg);
+    } else if (wpt_test) { // WPT test harness
+      wpt_test(function () {
+        wpt_assert_true(success, msg);
+      }, "WebGL test #" + (WPT_TEST_ID++) + ": " + msg);
+    }
   }
-}
+}())
 
 function notifyFinishedToHarness() {
   if (window.parent.webglTestHarness) {
@@ -92,7 +106,7 @@
       window.console.log(msg);
 }
 
-var _jsTestPreVerboseLogging = false;
+var _jsTestPreVerboseLogging = true;
 
 function enableJSTestPreVerboseLogging()
 {
@@ -105,31 +119,18 @@
     if (msg === undefined) {
       msg = document.title;
     }
-    // For MSIE 6 compatibility
-    var span = document.createElement("span");
-    span.innerHTML = '<p>' + msg + '</p><p>On success, you will see a series of "<span class="pass">PASS</span>" messages, followed by "<span class="pass">TEST COMPLETE</span>".</p>';
-    var description = document.getElementById("description");
-    if (description.firstChild)
-        description.replaceChild(span, description.firstChild);
-    else
-        description.appendChild(span);
-    if (_jsTestPreVerboseLogging) {
-        _logToConsole(msg);
-    }
+    _logToConsole("DESCRIPTION: " + msg);
 }
 
 function _addSpan(contents)
 {
-    var span = document.createElement("span");
-    document.getElementById("console").appendChild(span); // insert it first so XHTML knows the namespace
-    span.innerHTML = contents + '<br />';
 }
 
 function debug(msg)
 {
     _addSpan(msg);
     if (_jsTestPreVerboseLogging) {
-	_logToConsole(msg);
+        _logToConsole(msg);
     }
 }
 
@@ -143,7 +144,7 @@
     reportTestResultsToHarness(true, msg);
     _addSpan('<span><span class="pass">PASS</span> ' + escapeHTML(msg) + '</span>');
     if (_jsTestPreVerboseLogging) {
-	_logToConsole('PASS ' + msg);
+        _logToConsole('PASS ' + msg);
     }
 }
 
