--- conformance/more/unit.js	2015-06-18 23:26:41.085626000 +0200
+++ ../conformance-1.0.3/conformance/more/unit.js	2015-12-30 19:46:34.570636491 +0100
@@ -892,9 +892,14 @@
 0x809D
 ];
 
+var WPT_TEST_ID = 0;
 function reportTestResultsToHarness(success, msg) {
   if (window.parent.webglTestHarness) {
     window.parent.webglTestHarness.reportResults(window.location.pathname, success, msg);
+  } else if (window.test) { // WPT test harness
+    test(function () {
+      assert_true(success, msg);
+    }, "WebGL test #" + (WPT_TEST_ID++) + ": " + msg);
   }
 }
 
