--- WebIDL.py
+++ WebIDL.py
@@ -1744,7 +1744,8 @@
                   identifier == "ProbablyShortLivingWrapper" or
                   identifier == "LegacyUnenumerableNamedProperties" or
                   identifier == "RunConstructorInCallerCompartment" or
-                  identifier == "NonOrdinaryGetPrototypeOf"):
+                  identifier == "NonOrdinaryGetPrototypeOf" or
+                  identifier == "Abstract"):
                 # Known extended attributes that do not take values
                 if not attr.noArguments():
                     raise WebIDLError("[%s] must take no arguments" % identifier,
