--- WebIDL.py
+++ WebIDL.py
@@ -1416,7 +1416,8 @@
                   identifier == "LegacyEventInit" or
                   identifier == "ProbablyShortLivingObject" or
                   identifier == "LegacyUnenumerableNamedProperties" or
-                  identifier == "NonOrdinaryGetPrototypeOf"):
+                  identifier == "NonOrdinaryGetPrototypeOf" or
+                  identifier == "Abstract"):
                 # Known extended attributes that do not take values
                 if not attr.noArguments():
                     raise WebIDLError("[%s] must take no arguments" % identifier,
