--- WebIDL.py
+++ WebIDL.py
@@ -1768,7 +1768,8 @@ class IDLInterface(IDLInterfaceOrNamespace):
                   identifier == "LegacyUnenumerableNamedProperties" or
                   identifier == "RunConstructorInCallerCompartment" or
                   identifier == "WantsEventListenerHooks" or
-                  identifier == "Serializable"):
+                  identifier == "Serializable" or
+                  identifier == "Abstract"):
                 # Known extended attributes that do not take values
                 if not attr.noArguments():
                     raise WebIDLError("[%s] must take no arguments" % identifier,
