--- WebIDL.py
+++ WebIDL.py
@@ -7123,7 +7123,8 @@ class Parser(Tokenizer):
             self.parser = yacc.yacc(module=self,
                                     outputdir=outputdir,
                                     tabmodule='webidlyacc',
-                                    errorlog=logger
+                                    errorlog=logger,
+                                    debug=False
                                     # Pickling the grammar is a speedup in
                                     # some cases (older Python?) but a
                                     # significant slowdown in others.
