--- WebIDL.py
+++ WebIDL.py
@@ -1988,6 +1988,7 @@ class IDLInterface(IDLInterfaceOrNamespace):
                 or identifier == "WantsEventListenerHooks"
                 or identifier == "Serializable"
                 or identifier == "Abstract"
+                or identifier == "Inline"
             ):
                 # Known extended attributes that do not take values
                 if not attr.noArguments():
