--- WebIDL.py
+++ WebIDL.py
@@ -4570,8 +4570,7 @@ class IDLArgument(IDLObjectWithIdentifier):
 
         if ((self.type.isDictionary() or
              self.type.isUnion() and self.type.unroll().hasDictionaryType()) and
-            self.optional and not self.defaultValue and not self.variadic and
-            not self.dictionaryMember):
+            self.optional and not self.defaultValue and not self.variadic):
             # Default optional non-variadic dictionary arguments to null,
             # for simplicity, so the codegen doesn't have to special-case this.
             self.defaultValue = IDLNullValue(self.location)
